snafu/benchmarks/uperf.py

```diff
@@ -0,0 +1,242 @@
+        # We assume message_size=write_message_size to prevent breaking dependant implementations
+
+        return UperfStdout(
+            results=tuple(
+                RawUperfResult(timestamp=float(r[0]), bytes=int(r[1]), ops=int(r[2])) for r in results
+            ),
+            config=UperfConfig(
+                test_type=test_type,
+                protocol=protocol,
+                message_size=int(wsize),
+                read_message_size=int(rsize),
+                num_threads=int(nthr),
+                duration=len(results),
+            ),
+        )
+
+    def get_results_from_stdout(self, stdout: UperfStdout) -> List[UperfResultData]:
+        """Return list of results given raw uperf stdout."""
+
+        processed: List[BenchmarkResult] = []
+        prev_bytes: int = 0
+        prev_ops: int = 0
+        prev_timestamp: float = 0.0
+        bytes: int = 0
+        ops: int = 0
+        timestamp: float = 0.0
+        norm_ops: int = 0
+        norm_ltcy: float = 0.0
+
+        for result in stdout["results"]:
+            timestamp, bytes, ops = result["timestamp"], result["bytes"], result["ops"]
+
+            norm_ops = ops - prev_ops
+            if norm_ops == 0:
+                norm_ltcy = 0.0
+            else:
+                norm_ltcy = ((timestamp - prev_timestamp) / norm_ops) * 1000
+
+            datapoint = UperfResultData(
+                uperf_ts=datetime.datetime.fromtimestamp(int(timestamp) / 1000),
+                bytes=bytes,
+                norm_byte=bytes - prev_bytes,
+                ops=ops,
+                norm_ops=norm_ops,
+                norm_ltcy=norm_ltcy,
+            )
+
+            processed.append(datapoint)
+            prev_timestamp, prev_bytes, prev_ops = timestamp, bytes, ops
+
+        return processed
+
+    def setup(self) -> bool:
+        """Parse config and check that workload file exists."""
+        self.config.parse_args()
+
+        if isinstance(self.config.density_range, str):
+            self.config.config.density_range = [
+                int(x) for x in self.config.config.density_range.split("-") if x != ""
+            ]
+        if isinstance(self.config.node_range, str):
+            self.config.config.node_range = [
+                int(x) for x in self.config.config.node_range.split("-") if x != ""
+            ]
+        if not check_file(self.config.workload):
+            self.logger.critical(f"Unable to read workload file located at {self.config.workload}")
+            return False
+
+        return True
+
+    def run(self) -> Iterable[BenchmarkResult]:
+        """
+        Run uperf benchmark ``self.config.sample`` number of times.
+
+        Returns immediately if a sample fails. Will attempt to Uperf run three times for each sample.
+        """
+
+        self.logger.info("Running setup tasks.")
+        if not self.setup():
+            self.logger.critical(f"Something went wrong during setup, refusing to run.")
+            return
```

Comment:
This code would have to be run in every ``run`` method. Maybe remove from the run method, and create a new method called ``get_results`` or something that runs ``setup`` and then ``run``.

Reply:
You could move the logic in `setup` into the Class `__init__` method